ui: Auto-zoom map to fit visible content

mrtracy opened the pull request:
The cluster map now auto-zooms to the set of visible information
whenever the user navigates to a new locality, or whenever new locality
information becomes available for the locality currently being viewed.

Fixes #23513

Release note: None

![screen shot 2018-03-14 at 5 06 40 pm](https://user-images.githubusercontent.com/6969858/37430947-27947586-27aa-11e8-953e-d2b3ef7843a3.png)

Commit:
ui: Auto-zoom map to fit visible content

The cluster map now auto-zooms to the set of visible information
whenever the user navigates to a new locality, or whenever new locality
information becomes available for the locality currently being viewed.

Fixes #23513

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/23866)
<!-- Reviewable:end -->

vilterp commented:
Verified locally whoooooo

vilterp commented on pkg/ui/ccl/src/views/clusterviz/containers/map/mapLayout.tsx:
```diff
@@ -29,15 +31,35 @@ interface MapLayoutProps {
+        };
+      });
+    },
+  );
```
nice. Eventually let's move this up to one of the selectors files so we can use it from the locality tree debug page and elsewhere.

couchand commented:
Thanks for taking care of this @mrtracy.  In general I would prefer if nontrivial PRs were open for more than a few hours to make sure there's enough time for them to be reviewed.

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/mapLayout.tsx, line 49 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23866#-L7e2cYn6HAeYtCqdyW9:-L7e2cYn6HAeYtCqdyWA:b-viq68c) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce7167b9847a2b382b18d1a43c66e61e1a899c5/pkg/ui/ccl/src/views/clusterviz/containers/map/mapLayout.tsx#L49)):*
> ```XML
>   maxLatitude = 80;
> 
>   localityLocations = createSelector(
> ```

i don't think `createSelector` is warranted here.  can this just be a method, since it will have access to `this.props` anyway?

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/mapLayout.tsx, line 61 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23866#-L7aKl0R2b_Ir-5Xr9vP-r1-61:-L7e0PmoFty8G2bRzVDw:b-ypwjlz) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce7167b9847a2b382b18d1a43c66e61e1a899c5/pkg/ui/ccl/src/views/clusterviz/containers/map/mapLayout.tsx#L61)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

nice. Eventually let's move this up to one of the selectors files so we can use it from the locality tree debug page and elsewhere.
</blockquote></details>

this isn't properly a selector, since it doesn't operate on the redux state...

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/mapLayout.tsx, line 119 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23866#-L7e08FMTvUGVPN_E1Je:-L7e08FN_5diiWasM0uf:brpix6b) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce7167b9847a2b382b18d1a43c66e61e1a899c5/pkg/ui/ccl/src/views/clusterviz/containers/map/mapLayout.tsx#L119)):*
> ```XML
>     // locality tree OR that new data has been added to the currently visible
>     // locality.
>     if (_.isEqual(localityLocations, prevLocalityLocations)) {
> ```

it seems like it would be preferable if this check didn't fail if the localities were reordered.  it's not really a big deal since they probably won't be, but that would be an annoying bug to track down.  perhaps just keeping the tier structure rather than a flat list would be better?

fyi an outstanding refactor we've talked about is making the `LocalityTree` generic on the data stored for each locality, so that running `findOrCalculateLocation` on a given locality converts to a new data structure that has location data attached.  but that's almost certainly out of scope for 2.0

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/mapLayout.tsx, line 149 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23866#-L7e2HcX_isvZ75tLfKL:-L7e2HcX_isvZ75tLfKM:bvt3pit) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce7167b9847a2b382b18d1a43c66e61e1a899c5/pkg/ui/ccl/src/views/clusterviz/containers/map/mapLayout.tsx#L149)):*
> ```XML
>     this.updateZoomState(zoomTransform);
>     this.setState({ zoomTransform });
>     this.rezoomToLocalities();
> ```

seems like it would be cleaner to only zoom once here

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/zoom.ts, line 15 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23866#-L7e3H4myznbBu596Jgk:-L7e3H4myznbBu596Jgl:bwul6a5) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce7167b9847a2b382b18d1a43c66e61e1a899c5/pkg/ui/ccl/src/views/clusterviz/containers/map/zoom.ts#L15)):*
> ```XML
> export class Box {
>   // Compute a minimum bounding box for a supplied collection of boxes.
>   static boundingBox(...boxes: Box[]): Box {
> ```

i was going to suggest a unit test for this method, but i guess this whole file is untested...

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/zoom.ts, line 28 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23866#-L7e3-7UdGb-hioCipij:-L7e3-7UdGb-hioCipik:b9w4crr) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce7167b9847a2b382b18d1a43c66e61e1a899c5/pkg/ui/ccl/src/views/clusterviz/containers/map/zoom.ts#L28)):*
> ```XML
>       right = Math.max(right, boxes[i].right());
>       left = Math.min(left, boxes[i].left());
>       bottom = Math.max(bottom, boxes[i].bottom());
> ```

you could use `d3.min`/`d3.max` for these and avoid the for loop + infinity initializers.

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/zoom.ts, line 103 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23866#-L7e3WFb7k8op3dv6EdI:-L7e3WFb7k8op3dv6EdJ:b-oxyp9x) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce7167b9847a2b382b18d1a43c66e61e1a899c5/pkg/ui/ccl/src/views/clusterviz/containers/map/zoom.ts#L103)):*
> ```XML
>     this._viewportSize = viewportSize;
>     this._scale = this.minScale();
>     this._translate =  vector.sub(vector.mult(this._viewportSize, 0.5), vector.mult(this._bounds.center(), this._scale));
> ```

can you provide some context around this change?  was it wrong before?  perhaps if there were/weren't test changes that would be more clear.

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/zoom.ts, line 142 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23866#-L7e3jfaKzNyn-rLjciR:-L7e3jfaKzNyn-rLjciS:b-scivll) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce7167b9847a2b382b18d1a43c66e61e1a899c5/pkg/ui/ccl/src/views/clusterviz/containers/map/zoom.ts#L142)):*
> ```XML
>   }
> 
>   // zoomedToBoundingBox returns a ZoomTransformer which has been adjusted to
> ```

s/zoomedToBoundingBox/zoomedToBox/

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/zoom.ts, line 143 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23866#-L7e3vQGyfUrG5amJGFl:-L7e3vQGyfUrG5amJGFm:bmrkfhz) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce7167b9847a2b382b18d1a43c66e61e1a899c5/pkg/ui/ccl/src/views/clusterviz/containers/map/zoom.ts#L143)):*
> ```XML
> 
>   // zoomedToBoundingBox returns a ZoomTransformer which has been adjusted to
>   // exactly fit and center the provided bounding box.
> ```

can you clarify this comment in the context of a box that's not the same aspect ratio as the viewport?

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/zoom.ts, line 155 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23866#-L7e4S8Ng3DVcVmhACWd:-L7e4S8Ng3DVcVmhACWe:bm95fc0) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce7167b9847a2b382b18d1a43c66e61e1a899c5/pkg/ui/ccl/src/views/clusterviz/containers/map/zoom.ts#L155)):*
> ```XML
>       this._viewportSize[1] / boundingSize[1],
>     );
>     newZoom._translate =  vector.sub(vector.mult(this._viewportSize, 0.5), vector.mult(bounding.center(), newZoom._scale));
> ```

it looks like there's DRY opportunities here with the constructor

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23866#-:-L7e4ZUi9dxMNpntFsLS:b690115)*
<!-- Sent from Reviewable.io -->

vilterp commented:
Whoops, maybe pulled the trigger a little too quick. #23883 is still out, so @mrtracy could fix these things there if desired.

mrtracy commented:
@couchand I've addressed your comments as part of #23883

---

Review status: all files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/mapLayout.tsx, line 49 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23866#-L7e2cYn6HAeYtCqdyW9:-L7frsk16MZWcg4inElC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce7167b9847a2b382b18d1a43c66e61e1a899c5/pkg/ui/ccl/src/views/clusterviz/containers/map/mapLayout.tsx#L49)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

i don't think `createSelector` is warranted here.  can this just be a method, since it will have access to `this.props` anyway?
</blockquote></details>

Done.

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/mapLayout.tsx, line 61 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23866#-L7aKl0R2b_Ir-5Xr9vP-r1-61:-L7frrJbCqWEFfsHldN3:b-f4ddum) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce7167b9847a2b382b18d1a43c66e61e1a899c5/pkg/ui/ccl/src/views/clusterviz/containers/map/mapLayout.tsx#L61)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

this isn't properly a selector, since it doesn't operate on the redux state...
</blockquote></details>

I've removed this.

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/mapLayout.tsx, line 119 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23866#-L7e08FMTvUGVPN_E1Je:-L7frfW73mr3GP5ylKFz:b-jr26tt) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce7167b9847a2b382b18d1a43c66e61e1a899c5/pkg/ui/ccl/src/views/clusterviz/containers/map/mapLayout.tsx#L119)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

it seems like it would be preferable if this check didn't fail if the localities were reordered.  it's not really a big deal since they probably won't be, but that would be an annoying bug to track down.  perhaps just keeping the tier structure rather than a flat list would be better?

fyi an outstanding refactor we've talked about is making the `LocalityTree` generic on the data stored for each locality, so that running `findOrCalculateLocation` on a given locality converts to a new data structure that has location data attached.  but that's almost certainly out of scope for 2.0
</blockquote></details>

We can't check the entire tier structure because it includes metrics data and that changes.  The locations themselves seemed like the most relevant thing to compare, as they are the only thing needed by the zoom.  I'm not concerned about random sort-order changes, I don't think that's something that will happen.

Even if it did, I think this would actually be a pretty easy bug to catch, as the _.isEqual check is pretty conspicuous and if its unexpectedly getting bypassed these are easy enough to difference.

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/mapLayout.tsx, line 149 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23866#-L7e2HcX_isvZ75tLfKL:-L7fjOQgnZU35PQGXpMY:b-ts7c0q) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce7167b9847a2b382b18d1a43c66e61e1a899c5/pkg/ui/ccl/src/views/clusterviz/containers/map/mapLayout.tsx#L149)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

seems like it would be cleaner to only zoom once here
</blockquote></details>

I've refactored this so I don't need to call updateZoomState here.

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/zoom.ts, line 15 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23866#-L7e3H4myznbBu596Jgk:-L7fjABF7mGVpL8b348o:bh2nc5t) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce7167b9847a2b382b18d1a43c66e61e1a899c5/pkg/ui/ccl/src/views/clusterviz/containers/map/zoom.ts#L15)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

i was going to suggest a unit test for this method, but i guess this whole file is untested...
</blockquote></details>

Was my thought as well.

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/zoom.ts, line 28 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23866#-L7e3-7UdGb-hioCipij:-L7fE1gU9mnUT_857m0Z:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce7167b9847a2b382b18d1a43c66e61e1a899c5/pkg/ui/ccl/src/views/clusterviz/containers/map/zoom.ts#L28)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

you could use `d3.min`/`d3.max` for these and avoid the for loop + infinity initializers.
</blockquote></details>

Done.

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/zoom.ts, line 103 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23866#-L7e3WFb7k8op3dv6EdI:-L7fE24mA0a8PCz70Dgm:b-brv8ww) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce7167b9847a2b382b18d1a43c66e61e1a899c5/pkg/ui/ccl/src/views/clusterviz/containers/map/zoom.ts#L103)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

can you provide some context around this change?  was it wrong before?  perhaps if there were/weren't test changes that would be more clear.
</blockquote></details>

Yes, it was wrong before; the viewport vector should not have been modified by the scale, the translation is applied after zooming but the viewport size isn't scaled.

I've extracted this to a method "centerOnBox" and added some explanatory comments.

I think it appeared to "work" before because it was only called on constructor and "adjustZoom" was fixing it somewhat.

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/zoom.ts, line 142 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23866#-L7e3jfaKzNyn-rLjciR:-L7fBdli78AvxrJGFqda:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce7167b9847a2b382b18d1a43c66e61e1a899c5/pkg/ui/ccl/src/views/clusterviz/containers/map/zoom.ts#L142)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

s/zoomedToBoundingBox/zoomedToBox/
</blockquote></details>

Done.

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/zoom.ts, line 143 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23866#-L7e3vQGyfUrG5amJGFl:-L7fBdNP0L9Y5Bgxe2ds:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce7167b9847a2b382b18d1a43c66e61e1a899c5/pkg/ui/ccl/src/views/clusterviz/containers/map/zoom.ts#L143)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

can you clarify this comment in the context of a box that's not the same aspect ratio as the viewport?
</blockquote></details>

Done.

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/zoom.ts, line 155 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23866#-L7e4S8Ng3DVcVmhACWd:-L7fDxFrCtHn77M2qvvg:bjsk61) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce7167b9847a2b382b18d1a43c66e61e1a899c5/pkg/ui/ccl/src/views/clusterviz/containers/map/zoom.ts#L155)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

it looks like there's DRY opportunities here with the constructor
</blockquote></details>

Extracted a private adjustment method "centerOnBox", added comments explaining the transformation.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23866#-:-L7fsPZVAt7lka3agXsg:be2fliu)*
<!-- Sent from Reviewable.io -->
